### ISSUE TITLE
HMR doesn't work with Fusebox

### ISSUE THREAD

**Issue by auchenberg** [microsoft/vscode-chrome-debug#459]
Follow up on .https://github.com/Microsoft/vscode-chrome-debug/issues/216#issuecomment-315626341 @andrew-w-ross

Looks like Fuse-Box's HMR is pushing a new sourcemap but it can't be found/parsed by the debugger, which causes breakpoints to be re-mapped at the wrong position.

**Steps to reproduce:**
1. Get https://github.com/andrew-w-ross/fuse-debug
2. Start via F5
3. Put a breakpoint here, https://github.com/andrew-w-ross/fuse-debug/blob/master/src/components/Adder.tsx#L19
4. Change the value being added and the break point will no longer break

**Verbose log**
```
From target: {"method":"Debugger.scriptParsed","params":{"scriptId":"43","url":"","startLine":0,"startColumn":0,"endLine":28,"endColumn":2,"executionContextId":1,"hash":"CA1491A0413B2B2610DA0706E9A61D4A3FA4A126","executionContextAuxData":{"isDefault":true,"frameId":"95818.1"},"isLiveEdit":false,"sourceMapURL":"Adder.js.map","hasSourceURL":false,"isModule":false,"length":823,"stackTrace":{"callFrames":[{"functionName":"","scriptId":"33","url":"http://localhost:4444/app.js","lineNumber":22176,"columnNumber":4339}]}}}
SourceMaps.getMapForGeneratedPath: Finding SourceMap for VM43 by URI: Adder.js.map and webRoot: /Users/auchenberg/development/fuse-debug/src
SourceMaps.loadSourceMapContents: Reading local sourcemap file from Adder.js.map
To client: {"seq":0,"type":"event","event":"script","body":{"reason":"new","script":{"id":3,"source":{"name":"VM43","path":"VM43","sourceReference":1003}}}}
SourceMaps.loadSourceMapContents: Could not read sourcemap file - ENOENT: no such file or directory, open 'Adder.js.map'
SourceMaps.loadSourceMapContents: Reading local sourcemap file from VM43.map
SourceMaps.loadSourceMapContents: Could not read sourcemap file - ENOENT: no such file or directory, open 'VM43.map'
```

**Comment by auchenberg** [comment]
Making a HTTP request to `http://localhost:4444/Adder.js.map` returns a HTTP 400

**Comment by roblourens** [comment]
Where is Adder.js.map actually? Do you know how chrome devtools finds it?

**Comment by auchenberg** [comment]
Adder.js.map` doesn't seem to exist, a new sourcemap is generated for `app.js.map`, but that isn't pushed over HMR. Chrome DevTools doesn't break after HMR either, so it's a general fuse-box problem, and I've opened an issue https://github.com/fuse-box/fuse-box/issues/673

**Comment by andrew-w-ross** [comment]
@auchenberg I'm attempting to make a fix for fuse-box/fuse-box#673. My version of fuse-box now points to the existing sourcemap path but that doesn't seem to work and now I'm a bit stumped.

Here are the steps to recreate it:

- Check out my fixed version of fuse-box `git clone -b fix-hmr-sourcemap https://github.com/andrew-w-ross/fuse-box.git ` 
- Do the install, build and link:
  - `cd fuse-box`
  - `npm i`
  - `npm run dist`
  - `npm link`
- Then check out my debug project `git clone -b fix-hmr-sourcemap https://github.com/andrew-w-ross/fuse-debug.git ` 
- Do the install and link:
  - `cd fuse-debug`
  - `npm i`
  - `npm link fuse-box`
- open the project with vs code and launch with the `Start Both` configuration. 
- add a break point to [the add function](https://github.com/andrew-w-ross/fuse-debug/blob/fix-hmr-sourcemap/src/components/Adder.tsx#L19)
- click the click me button and and the break point should hit
- update the value in Adder to some other number and save, wait for the page to hot reload and the breakpoint should go from red -> grey -> red again
- click the click me button and the breakpoint won't hit but the value updates correctly.

I suspect the issue is the new change is done in memory in so I don't see how it would pickup that the source has changed. If you want to debug the hot reload code it'll be in line 80 of dist/app.js of the webapp, ironically you'll need to turn off sourcemaps.

**Comment by auchenberg** [comment]
Breifly looked into this. 

This is the log from VS Code after a HMR change is pushed over the WS
```
From target: {"method":"Debugger.scriptParsed","params":{"scriptId":"34","url":"","startLine":0,"startColumn":0,"endLine":28,"endColumn":2,"executionContextId":1,"hash":"CA1491A0571173D51417A210E9A61D4A3FA4A126","executionContextAuxData":{"isDefault":true,"frameId":"18477.1"},"isLiveEdit":false,"sourceMapURL":"app.js.map","hasSourceURL":false,"isModule":false,"length":821,"stackTrace":{"callFrames":[{"functionName":"","scriptId":"33","url":"http://localhost:4444/app.js","lineNumber":22518,"columnNumber":26}]}}}
SourceMaps.getMapForGeneratedPath: Finding SourceMap for VM34 by URI: app.js.map and webRoot: /Users/auchenberg/development/fuse-debug
SourceMaps.loadSourceMapContents: Reading local sourcemap file from app.js.map
To client: {"seq":0,"type":"event","event":"script","body":{"reason":"new","script":{"id":2,"source":{"name":"VM34","path":"VM34","sourceReference":1002}}}}
SourceMaps.loadSourceMapContents: Could not read sourcemap file - ENOENT: no such file or directory, open 'app.js.map'
SourceMaps.loadSourceMapContents: Reading local sourcemap file from VM34.map
SourceMaps.loadSourceMapContents: Could not read sourcemap file - ENOENT: no such file or directory, open 'VM34.map'
```

VS Code log when the the script is loaded on a page load
```
From target: {"method":"Debugger.scriptParsed","params":{"scriptId":"33","url":"http://localhost:4444/app.js","startLine":0,"startColumn":0,"endLine":22556,"endColumn":31,"executionContextId":1,"hash":"34AF15C29C6C9508BABBD5721EC1D2309E547D26","executionContextAuxData":{"isDefault":true,"frameId":"18723.1"},"isLiveEdit":false,"sourceMapURL":"app.js.map","hasSourceURL":false,"isModule":false,"length":778185}}
Paths.scriptParsed: could not resolve http://localhost:4444/app.js to a file under webRoot: /Users/auchenberg/development/fuse-debug. It may be external or served directly from the server's memory (and that's OK).
SourceMaps.getMapForGeneratedPath: Finding SourceMap for http://localhost:4444/app.js by URI: app.js.map and webRoot: /Users/auchenberg/development/fuse-debug
SourceMaps.loadSourceMapContents: Downloading sourcemap file from http://localhost:4444/app.js.map
To client: {"seq":0,"type":"event","event":"script","body":{"reason":"new","script":{"id":1,"source":{"name":"app.js","path":"http://localhost:4444/app.js","sourceReference":1001}}}}
```

The difference seems to be that `Debugger.scriptParsed` has a absolute `url` property, and the HMR script doesn't. This could be the reason for why the sourcemap URL is resolved badly for the HMR update @roblourens 

I wonder what Webpack does in this case.

@andrew-w-ross The basic problem is that our debugger can't fetch the updated sourcemap, and therefore fails.

**Comment by roblourens** [comment]
The error message implies that missing the url could be the issue here, although I know that I worked on the "eval scripts with sourcemaps" case at one point.

**Comment by andrew-w-ross** [comment]
I've had a look at how webpack does it to try and see what it does that fuse-box doesn't to try get it working. But it doesn't seem to work either on HMR unless you use `eval-source-map` option with the webpack devtool.

I've made a simple project here https://github.com/andrew-w-ross/webpack-debug. The funny part is it does seem to work fine in chrome directly,

**Comment by auchenberg** [comment]
@andrew-w-ross Our debugger works in a vastly different way then Chrome DevTools, as we rely on a well-working sourcemap with working paths, as we map the served assets to the local file system (open in Code), where Chrome DevTools simply works on inlined sources, which means paths can be wrong.

Our recommendation for Webpack is to use `source-map` or `eval-source-map`, which is used by boilerplates like `angular-cli`, `create-react-app` or `react-boilerplate`.

**Comment by andrew-w-ross** [comment]
@auchenberg Ok I won't look at how chrome does this when investigating this issue then. Can this issue be reopened as there still seems to be a problem even though I got the souremap url fixed?

**Comment by auchenberg** [comment]
@andrew-w-ross The problem seem to be the way you evaluate your HMR push update in Chrome, which causes Chrome to trigger a `Debugger.scriptParsed` event with no `url`, which causes our debugger to fail composing the right path for the referenced sourcemap.

This works for Webpack's HMR, so I would take a look at Webpacks implementation and see how thet evaluate things to get the URL right.

**Comment by andrew-w-ross** [comment]
The way it evaluates the HMR update is pretty simple. It essentially does the followoing
```javascript
  var source = `console.log("Hello World");
//# sourceMappingURL=somefile.js.map`;

  var res = new Function("some", "module", "args", source);
  res(module, stuff);
```

I assume because the new script isn't evaluated from a newly added file that's why there is no url? So perhaps default to base url in that case.

**Comment by auchenberg** [comment]
We can't just make assumptions about a base-path, as scripts can be loaded from 3rd party domains, within iframes, etc. The referenced URls `Debugger.scriptParsed` are the guide.

I just debugged Webpack's HMR, and for HMR updates they push scripts with `url`, so they must evaluate the code differently.

```
Paths.scriptParsed: could not resolve http://localhost:8080/0.d81117af64c462991b2f.hot-update.js to a file under webRoot: /Users/auchenberg/development/webpack-debug. It may be external or served directly from the server's memory (and that's OK).
SourceMaps.getMapForGeneratedPath: Finding SourceMap for http://localhost:8080/0.d81117af64c462991b2f.hot-update.js by URI: 0.d81117af64c462991b2f.hot-update.js.map and webRoot: /Users/auchenberg/development/webpack-debug
```

I don't have a better suggestion for us to "figure out the path", unless fuse-box could adopt the same approach as webpack or include an absolute url to the sourcemap?

**Comment by andrew-w-ross** [comment]
Yeah the difference is webpack adds a script tag in the head with the patch and fuse-box uses the function constructor. After looking at the documentation for `Debugger.scriptParsed` url is optional so this might be an issue you run into again in the future. Why not fallback to the url of the script that evaluated it?

**Comment by auchenberg** [comment]
@roblourens How do we handle evaluated scripts today? I know there's been some parallel work going on for the Node adapter. We could extract a url from the first frame in `stackTrace` if `url` is missing, and use that as the basepath to resolve the referenced sourcemap, but I feels a bit dirty.

**Comment by roblourens** [comment]
If we can combine sourceMapURL with webRoot or `http://localhost`, then we should be able to handle that case. If not, then I don't see how we can. Using the url of the script that evaled the script seems likely to be wrong.

**Comment by auchenberg** [comment]
Looking at the log in https://github.com/Microsoft/vscode-chrome-debug/issues/459#issuecomment-317587468 we already seem to try to combine `sourceMapURL` + `webRoot` by trying to read the sourcemap from the file system, but it does't exist.

@andrew-w-ross Do you persist the generated sourcemap or does it only exist as a virtual file?

**Comment by andrew-w-ross** [comment]
The original sourcemap file is updated.

**Comment by auchenberg** [comment]
I assume the problem is that `webRoot` points to 'src`, but the sourcemap exists in `dist` or whatever output folder from `fuse-box` which causes the sourcemap not to be found.

@andrew-w-ross You could define `# sourceURL` in the script you are evaluating within your HMR mechanism, that way you can tell the runtime and therefore the debugger, what file this evaluated script should be perceived as.

I just tried the following

```
  var source = `console.log("Hello World");
//# sourceMappingURL=app.js.map
//# sourceURL=http://localhost:4444/app.js`;

  var res = new Function("module", "args", source);
  res();
```
and it triggers the correct event flow:
```
From target: {"method":"Debugger.scriptParsed","params":{"scriptId":"55","url":"http://localhost:4444/app.js","startLine":0,"startColumn":0,"endLine":5,"endColumn":2,"executionContextId":8,"hash":"56D04F61816FD32861EAEC93D7AA5E8C2BA83281","executionContextAuxData":{"isDefault":true,"frameId":"39207.1"},"isLiveEdit":false,"sourceMapURL":"app.js.map","hasSourceURL":true,"isModule":false,"length":137,"stackTrace":{"callFrames":[{"functionName":"","scriptId":"54","url":"http://localhost:4444/app2.js","lineNumber":4,"columnNumber":12}]}}}
Paths.scriptParsed: could not resolve http://localhost:4444/app.js to a file under webRoot: /Users/auchenberg/development/fuse-debug. It may be external or served directly from the server's memory (and that's OK).
SourceMaps.getMapForGeneratedPath: Finding SourceMap for http://localhost:4444/app.js by URI: app.js.map and webRoot: /Users/auchenberg/development/fuse-debug
SourceMaps.loadSourceMapContents: Downloading sourcemap file from http://localhost:4444/app.js.ma
```

**Comment by andrew-w-ross** [comment]
@auchenberg  I'll give this a try and let you know what happens. Thanks for your time.

**Comment by auchenberg** [comment]
@andrew-w-ross Looks like our debugger gets confused about the script being evaluated as it's a script without a url, which cases us to try to fetch `app.js.map` which fails, as it should have been `http://localhost:4444/app.js.map`

Regular script load:
```
← From target: {"method":"Debugger.scriptParsed","params":{"scriptId":"32","url":"http://localhost:4444/app.js","startLine":0,"startColumn":0,"endLine":22556,"endColumn":31,"executionContextId":1,"hash":"7FDC8713B1914C90252AD455B0500C0596E782C6","executionContextAuxData":{"isDefault":true,"frameId":"15773.1"},"isLiveEdit":false,"sourceMapURL":"app.js.map","hasSourceURL":false,"isModule":false,"length":778184}}
Paths.scriptParsed: could not resolve http://localhost:4444/app.js to a file under webRoot: /Users/auchenberg/development/fuse-debug. It may be external or served directly from the server's memory (and that's OK).
SourceMaps.getMapForGeneratedPath: Finding SourceMap for http://localhost:4444/app.js by URI: app.js.map and webRoot: /Users/auchenberg/development/fuse-debug
SourceMaps.loadSourceMapContents: Downloading sourcemap file from http://localhost:4444/app.js.map
```

HMR update:
```
From target: {"method":"Debugger.scriptParsed","params":{"scriptId":"33","url":"","startLine":0,"startColumn":0,"endLine":28,"endColumn":2,"executionContextId":1,"hash":"CA1491A07E7D1FF81417A210E9A61D4A3FA4A126","executionContextAuxData":{"isDefault":true,"frameId":"15531.1"},"isLiveEdit":false,"sourceMapURL":"app.js.map","hasSourceURL":false,"isModule":false,"length":821,"stackTrace":{"callFrames":[{"functionName":"","scriptId":"30","url":"http://localhost:4444/app.js","lineNumber":22518,"columnNumber":26}]}}}
SourceMaps.getMapForGeneratedPath: Finding SourceMap for VM33 by URI: app.js.map and webRoot: /Users/auchenberg/development/fuse-debug
SourceMaps.loadSourceMapContents: Reading local sourcemap file from app.js.map
SourceMaps.loadSourceMapContents: Could not read sourcemap file - ENOENT: no such file or directory, open 'app.js.map'
```

**Comment by connor4312** [comment]
It looks like the fix for this was applied in fuse-box 4.0. However this is still tagged under the `next` tag after several years, and trying to debug any app there bounces the sourcemap request back with an unhelpful `429 Upgrade Required` response.

```
> GET /app.js.map HTTP/1.1
> Host: 127.0.0.1:4444
> User-Agent: curl/7.70.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 426 Upgrade Required
< Content-Length: 16
< Content-Type: text/plain
< Date: Wed, 04 Nov 2020 00:52:25 GMT
< Connection: keep-alive
```

I'm not really able to tell what it's expecting from the client, and as fusebox is hardly used any more, I don't think it's worth the time to investigate any more (though I would accept PRs).
